src/lib.rs

```diff
@@ -227,30 +227,34 @@ impl<T: Clone + Hash + Debug> Mphf<T> {
     /// and the size of the datastructure representing the hash function. See the paper for details.
     /// `max_iters` - None to never stop trying to find a perfect hash (safe if no duplicates).
     pub fn new(gamma: f64, objects: &[T]) -> Mphf<T> {
+        assert!(gamma > 1.01);
         let mut bitvecs = Vec::new();
         let mut iter = 0;
-        let mut redo_keys = Vec::new();
 
-        assert!(gamma > 1.01);
-        while !redo_keys.is_empty() || iter == 0 {
-            // this scope structure is needed to allow keys to be
-            // the 'objects' reference on the first loop, and a reference
-            // to 'redo_keys' on subsequent iterations.
-            redo_keys = {
-                let keys = if iter == 0 { objects } else { &redo_keys };
-
-                let cx = Context::new(
-                    std::cmp::max(255, (gamma * keys.len() as f64) as usize),
-                    iter,
-                );
-
-                (&keys).iter().for_each(|v| cx.find_collisions(v));
-                let redo_keys_tmp = (&keys).iter().filter_map(|v| cx.filter(v)).collect();
-
-                bitvecs.push(cx.a);
-                redo_keys_tmp
-            };
+        let cx = Context::new(
```

Comment:
This obviously adds some code duplication, but I find it quite a bit easier to follow, and it makes the next step (commit) a good deal easier (I couldn't come up with another way to avoid cloning).